### PR TITLE
Save entries without triggering events

### PR DESCRIPTION
**Pull request by stephanbouman**
Closes https://github.com/statamic/ideas/issues/460

I implemented the silently method for chaining in your Entry classes. If you don't want to trigger events while saving
`$entry->silently->save()` so you won't end up with an infinite loop of triggered events.

### CODE REVIEW

**Comment by duncanmcclean** [comment]
Nice PR, would you mind resolving the issues highlighted by StyleCI? https://github.styleci.io/analyses/yvKGEG?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by edalzell** [comment]
I’d prefer it mirror Laravel’s `saveQuietly` name

**Comment by stephanbouman** [comment]
@edalzell I was thinking of that option. I wasn't sure if it is OK to change the save-methods parameters. Or add them to be more precise. Should i go for save and delete quietly()?

**Comment by edalzell** [comment]
> @edalzell I was thinking of that option. I wasn't sure if it is OK to change the save-methods parameters. Or add them to be more precise. Should i go for save and delete quietly()?

As long as it's not a breaking change:
```
function saveQuietly() {
....
save(true);

...

function save($quietly = false)
....
```

Something like that is how I'd do it.

**Comment by stephanbouman** [comment]
I deleted the silently method. Now the save and delete method both have a quietly variant:
`$entry->saveQuietly()` and `$entry->deleteQuietly()`

**Comment by edalzell** [comment]
I think there’s a dispatchIf method, but not sure which version of laravel it came in at.

**Comment by duncanmcclean** [comment]
> I think there’s a dispatchIf method, but not sure which version of laravel it came in at.

Looks like it was introduced in Laravel 7 which probably means it can't be used, unfortunately.

**Comment by stephanbouman** [comment]
> > I think there’s a dispatchIf method, but not sure which version of laravel it came in at.
> 
> Looks like it was introduced in Laravel 7 which probably means it can't be used, unfortunately, as Statamic supports Laravel 6 as well.
> 
> (sorry @stephanbouman, I know you just changed it..)

No problem. I'm not very used to contributing. I still need to set up the styleci-thing in phpstorm; that could save me a lot of commits.

**Review by jasonvarga — CHANGES_REQUESTED** [review]
Thanks for this.

Unfortunately this will cause issues for people who have overridden the `Entry` class.
If they have a `save` or `delete` method, they'll get warnings telling them the method signature needs to match the parent class. (You mentioned your concern for this earlier).

I'd say use a `withEvents` property, but use dedicated `saveQuietly` method.
Also, we'll need to re-enable events after it's done.

```php
protected $withEvents = true;

public function saveQuietly()
{
    $this->withEvents = false;

    $result = $this->save();

    $this->withEvents = true;

    return $result;
}
```

Lastly, do you really need this to happen on `delete`? I don't know if it's necessary.
I'd prefer not to add it unless there was an actual use case for it.

**Comment by stephanbouman** [comment]
@jasonvarga I decided to remove the `deleteQuietly` method, since it is kind of impossible to fall into a loop when you deleted an entry. I liked it to have the same method available for save and delete, but I guess nobody will actually use it.

I created the `enableEvents` en `disableEvents` methods so maybe in future, when parameters could change, the method will be still the same.

File: src/Entries/Entry.php [hunk 1]
```diff
@@ -665,6 +680,16 @@ public function values()
...
+    public function disableEvents()
+    {
+        $this->withEvents = false;
+    }
+
+    public function enableEvents()
+    {
+        $this->withEvents = true;
+    }
+
```
Comment (jasonvarga) [inline src/Entries/Entry.php:692]: I don't think these methods are necessary right now. Just change the property within `saveQuietly`.

File: src/Entries/Entry.php [hunk 2]
```diff
@@ -327,9 +342,9 @@ public function order()
         }
 
         return $this->structure()->in($this->locale())
-            ->flattenedPages()
-            ->map->reference()
-            ->flip()->get($this->id) + 1;
+                   ->flattenedPages()
+                   ->map->reference()
+                   ->flip()->get($this->id) + 1;
```
Comment (jasonvarga) [inline src/Entries/Entry.php:347]: Please undo this indentation change.

**Review by jasonvarga — CHANGES_REQUESTED** [review]

**Review by jasonvarga — APPROVED** [review]

**Comment by jasonvarga** [comment]
Thanks again!

**Pull request merged**